File: src/Rulesage.Shared/Repositories/Abstractions/ICommunityRepository.cs
using Rulesage.Common.Grammar.Ast;

namespace Rulesage.Shared.Repositories.Abstractions;

public interface ICommunityRepository : IDocumentRepository
{
    Task<IEnumerable<(CommunityExpr, float)>> FindOrderByCosineDistanceAsync(string contextCommunity, float[] queryVector, int skip, int take,
        CancellationToken cancellationToken = default);
    
    Task<bool> SaveAsync(IEnumerable<CommunityExpr> communities, CancellationToken cancellationToken = default);
}

File: src/Rulesage.Common/Repositories/Abstractions/ICommunityRepository.fs
namespace Rulesage.Common.Repositories.Abstractions

open System.Runtime.InteropServices
open System.Threading
open System.Threading.Tasks
open Rulesage.Common.Grammar.Ast

type ICommunityRepository =
    inherit IDocumentRepository

    abstract FindOrderByCosineDistanceAsync:
        contextCommunity: string *
        queryVector: float32[] *
        skip: int *
        take: int *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<(CommunityExpr * float32) seq>

    abstract SaveAsync:
        communities: CommunityExpr seq *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<int>
